src/main/java/jenkins/plugins/logstash/persistence/LogzioDao.java

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+    }
+
+
+    private JsonObject createLogLine(JSONObject jsonData, String logMsg) {
+        JsonObject logLine = new JsonObject();
+
+        logLine.addProperty("message", logMsg);
+        logLine.addProperty("@timestamp", ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
+
+        for (Object key : jsonData.keySet()) {
+            if (!key.equals("message")){
+                logLine.addProperty(key.toString(), jsonData.getString(key.toString()));
+            }
+        }
+
+        return logLine;
+    }
+
+    @Override
+    public JSONObject buildPayload(BuildData buildData, String jenkinsUrl, List<String> logLines) {
+        JSONObject payload = new JSONObject();
+        payload.put("message", logLines);
+        payload.put("source", "jenkins");
+        payload.put("source_host", jenkinsUrl);
+        payload.put("@buildTimestamp", buildData.getTimestamp());
+        payload.put("@version", 1);
+        // flatten build data
+        Map<String, Object> flattenJson = JsonFlattener.flattenAsMap(buildData.toString());
+        for (Map.Entry<String, Object> entry : flattenJson.entrySet()) {
+            String key = entry.getKey().replace('.','_');
+            Object value = entry.getValue();
+            payload.put(key, value);
+        }
+
+        return payload;
+    }
+
+    @Override
+    public String getDescription(){ return host; }
+
+    public String getHost(){ return host; }
+
+    public String getKey(){ return key; }
+
+    public String getType(){ return TYPE; }
+
+    public static class LogzioHttpsClient{
+        private static final int MAX_SIZE_IN_BYTES = 8 * 1024 * 1024;  // 8 MB
+        private static final int CONNECT_TIMEOUT = 10 * 1000;
+        private static final int SOCKET_TIMEOUT = 10 * 1000;
+        private final HttpsSyncSender logzioClient;
+        private List<FormattedLogMessage> messages;
+        private int size;
+        private final SenderStatusReporter reporter;
+
+        LogzioHttpsClient(String token, String listener, String type) throws LogzioParameterErrorException {
+            HttpsRequestConfiguration gzipHttpsRequestConfiguration = HttpsRequestConfiguration
+                    .builder()
+                    .setLogzioToken(token)
+                    .setLogzioType(type)
+                    .setLogzioListenerUrl(listener)
+                    .setSocketTimeout(SOCKET_TIMEOUT)
+                    .setConnectTimeout(CONNECT_TIMEOUT)
+                    .setCompressRequests(true)
+                    .build();
+            reporter = new Reporter();
+            logzioClient = new HttpsSyncSender(gzipHttpsRequestConfiguration, reporter);
+            messages = new ArrayList<>();
+            size = 0;
+        }
+
+        public void send(FormattedLogMessage log) throws LogzioServerErrorException {
+            messages.add(log);
+            size += log.getSize();
+            if (size > MAX_SIZE_IN_BYTES) {
```

Comment:
this looks fishy
shouldn't you be flushing *before* the size is over the max limit not immediately after it's reached?

Reply:
I agree but it saves adding more logic to the shipper. The listener max bulk size is 10MB, I took 8MB as my upper limit, assuming a general log size is around 1KB + the fact I use gzip compression before sending it. Maybe I can make it clearer by having a `LISTENER_MAX_SIZE_IN_BYTES = 10 * 1024 *1024` and then define `MAX_SIZE_IN_BYTES = 0.8 * LISTENER_MAX_SIZE_IN_BYTES`. What do you think?

Reply:
This doesn't look that much more complex
```
  if (size +  log.getSize(); > MAX_SIZE_IN_BYTES)  sendAndReset();
  messages.add(log);
  size += log.getSize();
```
you still need to call flush at the end